Source/Core/Core/PowerPC/Expression.cpp

```diff
@@ -0,0 +1,249 @@
+    {"read_s8", HostReadFunc<s8, u8>},
+    {"read_u16", HostReadFunc<u16>},
+    {"read_s16", HostReadFunc<s16, u16>},
+    {"read_u32", HostReadFunc<u32>},
+    {"read_s32", HostReadFunc<s32, u32>},
+    {"read_f32", HostReadFunc<float, u32>},
+    {"read_f64", HostReadFunc<double, u64>},
+    {"write_u8", HostWriteFunc<u8>},
+    {"write_u16", HostWriteFunc<u16>},
+    {"write_u32", HostWriteFunc<u32>},
+    {"write_f32", HostWriteFunc<float, u32>},
+    {"write_f64", HostWriteFunc<double, u64>},
+    {"u8", CastFunc<u8>},
+    {"s8", CastFunc<s8, u8>},
+    {"u16", CastFunc<u16>},
+    {"s16", CastFunc<s16, u16>},
+    {"u32", CastFunc<u32>},
+    {"s32", CastFunc<s32, u32>},
+    {},
+}};
+
+void ExprDeleter::operator()(expr* expression) const
+{
+  expr_destroy(expression, nullptr);
+}
+
+void ExprVarListDeleter::operator()(expr_var_list* vars) const
+{
+  // Free list elements
+  expr_destroy(nullptr, vars);
+  // Free list object
+  delete vars;
+}
+
+Expression::Expression(std::string_view text, ExprPointer ex, ExprVarListPointer vars)
+    : m_text(text), m_expr(std::move(ex)), m_vars(std::move(vars))
+{
+  for (auto* v = m_vars->head; v != nullptr; v = v->next)
+  {
+    const std::string_view name = v->name;
+    VarBinding bind;
+
+    if (name.length() >= 2 && name.length() <= 3)
+    {
+      if (name[0] == 'r' || name[0] == 'f')
+      {
+        char* end = nullptr;
+        const int index = std::strtol(name.data() + 1, &end, 10);
+        if (index >= 0 && index <= 31 && end == name.data() + name.length())
+        {
+          bind.type = name[0] == 'r' ? VarBindingType::GPR : VarBindingType::FPR;
+          bind.index = index;
+        }
+      }
+      else if (name == "lr")
+      {
+        bind.type = VarBindingType::SPR;
+        bind.index = SPR_LR;
+      }
+      else if (name == "ctr")
+      {
+        bind.type = VarBindingType::SPR;
+        bind.index = SPR_CTR;
+      }
+      else if (name == "pc")
+      {
+        bind.type = VarBindingType::PCtr;
+      }
+    }
+
+    m_binds.emplace_back(bind);
+  }
+}
+
+std::optional<Expression> Expression::TryParse(std::string_view text)
+{
+  ExprVarListPointer vars{new expr_var_list{}};
+  ExprPointer ex{expr_create(text.data(), text.length(), vars.get(), g_expr_funcs.data())};
+  if (!ex)
+    return std::nullopt;
+
+  return Expression{text, std::move(ex), std::move(vars)};
+}
+
+double Expression::Evaluate() const
+{
+  SynchronizeBindings(SynchronizeDirection::From);
+
+  double result = expr_eval(m_expr.get());
+
+  if (isnan(result))
+  {
+    Core::DisplayMessage("Breakpoint condition has returned NaN.", 2000);
+    NOTICE_LOG_FMT(MEMMAP, "Breakpoint condition returned NaN");
+  }
+
+  SynchronizeBindings(SynchronizeDirection::To);
+
+  return result;
+}
+
+void Expression::SynchronizeBindings(SynchronizeDirection dir) const
+{
+  auto bind = m_binds.begin();
+  for (auto* v = m_vars->head; v != nullptr; v = v->next, ++bind)
+  {
+    switch (bind->type)
```

Comment:
As far as I understand the code, `expr_var_list` also expands when users declare variables in the expression. The size of `m_vars` in this case will be bigger than `m_binds` and its iterator will go past its end with `++bind`.

We can also check in this function which variables are NaNs.

```C++
  auto bind = m_binds.begin();
  for (auto* v = m_vars->head; v != nullptr; v = v->next)
  {
    if (std::isnan(v->value))
    {
      // Warn the user about v->name being NaN
    }

    if (bind == m_binds.end())
      continue;

    switch (bind->type)
    {
      // ...
    }
    ++bind;
  }
```

Reply:
Non-binding variables are given VarBindingType::Zero: , so m_binds should equal m_vars.  My debug stepping test showed this on a simple test. Unless you have a specific case in mind I should test?

Your suggestion for NaN only checks if incoming register values are NaN ( can they be?).  If we want to check NaN in calculations, we have to do it in expr.h's expr_eval. I'll make a separate comment for that.

Reply:
Oh excuse me. After the expr eval, the m_vars have the final values applied to them and can be checked for NaN.  The only problem is a calculation without a variable.   f3 / f1 > 0. If f3 and f1 are both zero we get a NaN but there is no m_var attached to that, so we wouldn't know.  Do we need to know?

Reply:
> Non-binding variables are given VarBindingType::Zero: , so m_binds should equal m_vars. My debug stepping test showed this on a simple test. Unless you have a specific case in mind I should test?

Oh right, I thought the unbound variables weren't created by `expr_create` but by the `expr_eval` function. Whereas in reality they are created by `expr_create` and evaluated by `expr_eval`. That makes sense, feel free to disregard my comment about the past the end iterator.

> Your suggestion for NaN only checks if incoming register values are NaN ( can they be?). If we want to check NaN in calculations, we have to do it in expr.h's expr_eval.

Fair point, they can but only for FPRs and checking `SynchronizeDirection::From` doesn't make more sense. However, checking for NaN against non-FPR registers and variables during `SynchronizeDirection::To` does make sense since we're casting from `double`.

> The only problem is a calculation without a variable. f3 / f1 > 0. If f3 and f1 are both zero we get a NaN but there is no m_var attached to that, so we wouldn't know.

Sadly, there isn't much we can do in such case. I don't think tweaking the whole `expr.h` is the ideal way to go as you'd need to check for each operand. Plus, you don't know which values were intended to be floating-point and which weren't. 

> We could tell users to assign variables to test for NaN errors and we could also print the (double) value of any variable + the variable name to the log.

That sounds like a nice idea, especially for debugging expression output. That will also encourage users to use variables if they can see their return value.

Just to sum up my point of view regarding NaN. I think having NaNs in floating-point operations is fine and to be expected. But having them in _(supposedly)_ integer operations is most likely unexpected. This is a side-effect of everything being a `double`. That's why having a way to tell the user what's going on and what went wrong is meaningful. If we can't address the NaN issue completely, as long as users are aware of it, I would be fine with it.

Reply:
> However, checking for NaN against non-FPR registers and variables during `SynchronizeDirection::To` does make sense since we're casting from `double`.

NaN writes zeroes to a gpr. Out of range values seem to write somewhat random data.   Writing to registers was really just tacked on because it was easy to do.  The current logging will catch the intended output value and NaNs will be reported.  I'm not sure I want to work on `SynchronizeDirection::To` much, as it feels outside the scope of what the PR should be about. I wouldn't mind removing it.